TESTING.md

```diff
@@ -0,0 +1,274 @@
+
+**Good**:
+
+```ts
+describe('SomeComponent', () => {
+  it('should render the "text" prop as the text', async () => {
+    const {getByTestId} = render(<SomeComponent data-testid="test" text="foo" />);
+
+    const component = await getByTestId('test');
+    expect(component).toHaveTextContent('foo');
+  });
+
+  it('should render the "text" prop as an aria-label for accessibility', async () => {
+    const {getByTestId} = render(<SomeComponent data-testid="test" text="foo" />);
+
+    const component = await getByTestId('test');
+    expect(component).toHaveAttribute('aria-label', 'foo');
+  });
+});
+```
+
+This series of specifications give more context to the test. It is more code, but it is more clear
+what the intent of each specification is. If one fails, it will be easier to diagnose what and why.
+The assertions also make use of `jest-dom` which gives more context into _how_ it fails. Instead of
+a generic assertion, the CLI will output something like
+`Expected <div data-testid="test" aria-label="bar">bar</div> to have an "aria-label" of "foo" but received "bar"`.
+The specification should tell us which spec failed and the assertion gives insight into how.
+
+## Unit tests
+
+Canvas Kit uses [Jest](https://jestjs.io/),
+[React Testing Library](https://testing-library.com/docs/react-testing-library/intro), and
+[jest-dom](https://github.com/testing-library/jest-dom) which have a nice API for testing React
+components. Unit tests should test inputs and outputs of utility functions or props and output of
+components. For example, if documentation says that a component will pass extra props to a container
+component, there should be a specification that verifies this claim. If a component's documentation
+states that children will be rendered, a test should verify the children exist within the containing
+component.
+
+Good unit tests avoid too much implementation detail. For example, if a component states it will
+render the children prop, the documentation and specifications shouldn't state where unless it is
+specific to the functionality of the component. Tests here should define the API surface area that
+will be guaranteed. The more implementation details, the more difficult the tests are to maintain.
+
+**Bad**:
+
+```ts
+const {container} = render(
+  <SomeComponent>
+    <div data-testid="test">Test</div>
+  </SomeComponent>
+);
+
+expect(container.children[0].children[0].textContent).toEqual('Test');
+```
+
+This test encodes the DOM structure into it. Any changes to the source code will require a change to
+the test making the test tightly coupled and fragile.
+
+**Good**:
+
+```ts
+const {getByTestId} = render(
+  <SomeComponent data-testid="container">
+    <div data-testid="test">Test</div>
+  </SomeComponent>
+);
+
+const component = await getByTestId('container');
+const child = await getByTestId('test');
+
+expect(component).toContainElement(child);
+```
+
+The first test has the DOM structure encoded in it. The second
```

Comment:
Looks like you sneezed and hit delete in your editor here. Bless you.

Reply:
lol. I'll finish my thought

Reply:
This line seems redundant. The first sentence exists under the first example already